exporter/internal/queue/sized_elements_channel.go

```diff
@@ -0,0 +1,109 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package queue // import "go.opentelemetry.io/collector/exporter/internal/queue"
+
+import "sync/atomic"
+
+// sizedElementsChannel is a channel wrapper for sized elements with a capacity set to a total size of all the elements.
+// The channel will accept elements until the total size of the elements reaches the capacity.
+type sizedElementsChannel[T any] struct {
+	used *atomic.Int64
+
+	// We need to store the capacity in a separate field because the capacity of the channel can be higher.
+	// It happens when we restore a persistent queue from a disk that is bigger than the pre-configured capacity.
+	cap int64
+	ch  chan T
+}
+
+type sizedElementsChannelOption[T any] func(*sizedElementsChannel[T])
+
+// withPreloadElements puts the elements into the queue with the given size. It's used by the persistent queue to
+// initialize the queue with the elements recovered from the disk.
+func withPreloadElements[T any](els []T, totalSize int64) sizedElementsChannelOption[T] {
+	return func(vcq *sizedElementsChannel[T]) {
+		vcq.used.Store(totalSize)
+		if cap(vcq.ch) < len(els) {
+			vcq.ch = make(chan T, len(els))
+		}
+		for _, el := range els {
+			vcq.ch <- el
+		}
+	}
+}
+
+// newSizedElementsChannel creates a sized elements channel. Each element is assigned a size by the provided sizer.
+// chanCapacity is the capacity of the underlying channel which usually should be equal to the capacity of the queue to
+// avoid blocking the producer.
+func newSizedElementsChannel[T any](capacity int, opts ...sizedElementsChannelOption[T]) *sizedElementsChannel[T] {
+	sech := &sizedElementsChannel[T]{
+		used: &atomic.Int64{},
+		cap:  int64(capacity),
+		ch:   make(chan T, capacity),
```

Comment:
Useless allocation if this gets re-initialized in the options.

Reply:
Do you think we should avoid that? The override in the option happens under pretty rare conditions. We can avoid that but it'll add a bit of extra code after applying the options
```
	if sech.ch == nil {
		sech.ch = make(chan T, capacity)
	}
```

Reply:
I applied this